### PR TITLE
docs: overriding behavior for common labels

### LINKED ISSUE #4549: Common labels override labels from podMetadata
**What happened?**

When we updated prometheus-operator from ~v0.2.4~ v0.43.2 to v0.53.1 we lost some metrics because one Prometheus instance was not able to do service discovery anymore. This happened because we have some network policies in place that select pods based on the `app.kubernetes.io/instance` label. We used to define this label in the Prometheus `podMetadata` section, but now the label is always the name of the Prometheus resource.

**Did you expect to see something different?**

I expected that the Prometheus pods have the same `app.kubernetes.io/instance` label that is defined in `podMetadata`.

**How to reproduce it (as minimally and precisely as possible)**:

Create a Prometheus:

```yaml
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  labels:
    app.kubernetes.io/instance: owning-app
  name: prometheus
spec:
  podMetadata:
    labels:
      app.kubernetes.io/instance: owning-app
```

And expect a StatefulSet like this:

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/instance: owning-app
  name: prometheus-prometheus
spec:
  selector:
    matchLabels:
      app.kubernetes.io/instance: owning-app
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: owning-app
```

But get a StatefulSet like this:

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app.kubernetes.io/instance: owning-app
  name: prometheus-prometheus
spec:
  selector:
    matchLabels:
      app.kubernetes.io/instance: prometheus
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: prometheus
```

**Environment**

* Prometheus Operator version:

    `v0.53.1`

**Anything else we need to know?**:

I think this is due to #3841 where some common labels have been introduced. These common labels have a higher precedence than the labels from `podMetadata` and I think it should be the other way round, no?

### PR DESCRIPTION
**Pull request by rexagod**
## Description

Document common (reserved) labels in:
* `alertmanager`
* `prometheus`
* `thanos-ruler`

Fixes #4549.

## Type of change

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

Document overriding behavior for common (reserved) labels.

```release-note
- Document overriding behavior in common (reserved) labels for `alertmanager`, `prometheus`, and `thanos-ruler.
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 1]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
 // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
 // +k8s:openapi-gen=true
 type AlertmanagerSpec struct {
-	// PodMetadata configures Labels and Annotations which are propagated to the alertmanager pods.
+	// PodMetadata configures Labels and Annotations which are propagated to the alertmanager pods,
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:62]:
```suggestion
	// PodMetadata configures Labels and Annotations which are propagated to the Alertmanager pods,
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 2]
```diff
@@ -21,9 +21,8 @@ import (
 )
 
 const (
-	AlertmanagersKind   = "Alertmanager"
-	AlertmanagerName    = "alertmanagers"
-	AlertManagerKindKey = "alertmanager"
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:26]: not sure why you removed this const?

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 3]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
 // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
 // +k8s:openapi-gen=true
 type AlertmanagerSpec struct {
-	// PodMetadata configures Labels and Annotations which are propagated to the alertmanager pods.
+	// PodMetadata configures Labels and Annotations which are propagated to the alertmanager pods,
+	// out of which, the following are reserved and cannot be overridden:
+	// * "alertmanager": This will be set to the name of the Alertmanager instance.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:64]: (nit)
```suggestion
	// * "alertmanager" label, set to the name of the Alertmanager instance.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 4]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// * "app.kubernetes.io/instance": This will be set to the name of the Alertmanager instance.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:65]: (nit)
```suggestion
	// * "app.kubernetes.io/instance" label,  set to the name of the Alertmanager instance.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 5]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// * "app.kubernetes.io/managed-by": This will be set to "prometheus-operator" for all alertmanager pods.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:66]: (nit)
```suggestion
	// * "app.kubernetes.io/managed-by" label, set to "prometheus-operator".
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 6]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// * "app.kubernetes.io/name": This will be set to "alertmanager" for all alertmanager pods.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:67]:
```suggestion
	// * "app.kubernetes.io/name" label, set to "alertmanager".
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 7]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// * "app.kubernetes.io/version": This label is set to the Alertmanager version for all pods.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:68]:
```suggestion
	// * "app.kubernetes.io/version" label, set to the Alertmanager's version.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 8]
```diff
@@ -60,7 +59,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// * "kubectl.kubernetes.io/default-container": This annotation is managed by the operator and it will be set to "alertmanager".
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:69]:
```suggestion
	// * "kubectl.kubernetes.io/default-container" annotation, set to "alertmanager".
```

**Review by simonpasquier — COMMENTED** [review]
Thanks! I've got only some minor comments. I've focused on Alertmanager but if you agree with my suggestions, the same applies to Prometheus and ThanosRuler.

File: pkg/apis/monitoring/v1/alertmanager_types.go (hunk 2 above)
Comment (rexagod) [reply pkg/apis/monitoring/v1/alertmanager_types.go:26]: I think this isn't being used anywhere, I just searched again in the codebase at it still seems to be the case.

Comment (simonpasquier) [reply pkg/apis/monitoring/v1/alertmanager_types.go:26]: This is part of the API, we can't remove it (e.g. it may be used by external clients).

Comment (ArthurSens) [reply pkg/apis/monitoring/v1/alertmanager_types.go:26]: Related to https://github.com/prometheus-operator/prometheus-operator/pull/5772

File: pkg/apis/monitoring/v1/thanos_types.go [hunk 9]
```diff
@@ -68,7 +68,13 @@ type ThanosRulerList struct {
 type ThanosRulerSpec struct {
 	// Version of Thanos to be deployed.
 	Version string `json:"version,omitempty"`
-	// PodMetadata contains Labels and Annotations gets propagated to the thanos ruler pods.
+	// PodMetadata contains Labels and Annotations gets propagated to the ThanosRuler pods,
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/thanos_types.go:71]:
```suggestion
	// PodMetadata configures labels and annotations which are propagated to the ThanosRuler pods.
```

File: pkg/apis/monitoring/v1/thanos_types.go [hunk 10]
```diff
@@ -68,7 +68,13 @@ type ThanosRulerList struct {
...
+	// out of which, the following are reserved and cannot be overridden:
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/thanos_types.go:72]:
```suggestion
	//
	// The following items are reserved and cannot be overridden:
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 11]
```diff
@@ -59,7 +59,16 @@ func (l *Prometheus) GetStatus() PrometheusStatus {
 // CommonPrometheusFields are the options available to both the Prometheus server and agent.
 // +k8s:deepcopy-gen=true
 type CommonPrometheusFields struct {
-	// PodMetadata configures labels and annotations which are propagated to the Prometheus pods.
+	// PodMetadata configures labels and annotations which are propagated to the Prometheus pods,
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:62]:
```suggestion
	// PodMetadata configures labels and annotations which are propagated to the Prometheus pods.
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 12]
```diff
@@ -59,7 +59,16 @@ func (l *Prometheus) GetStatus() PrometheusStatus {
...
+	// out of which, the following are reserved and cannot be overridden:
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:63]:
```suggestion
	//
	// The following items are reserved and cannot be overridden:
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 13]
```diff
@@ -60,7 +60,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
 // https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
 // +k8s:openapi-gen=true
 type AlertmanagerSpec struct {
-	// PodMetadata configures Labels and Annotations which are propagated to the alertmanager pods.
+	// PodMetadata configures Labels and Annotations which are propagated to the Alertmanager pods,
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:63]:
```suggestion
	// PodMetadata configures labels and annotations which are propagated to the Alertmanager pods.
```

File: pkg/apis/monitoring/v1/alertmanager_types.go [hunk 14]
```diff
@@ -60,7 +60,14 @@ func (l *Alertmanager) DeepCopyObject() runtime.Object {
...
+	// out of which, the following are reserved and cannot be overridden:
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/alertmanager_types.go:64]:
```suggestion
	//
	// The following items are reserved and cannot be overridden:
```

**Review by simonpasquier — COMMENTED** [review]
Just a few nits :)

**Comment by rexagod** [comment]
I'm encountering errors on `make --always-make format generate`, I'll try getting the binary manually from source.

```
Downloading shellcheck
tar: shellcheck-stable/shellcheck: Not found in archive
tar: Error exit delayed from previous errors.
make: *** [/Users/rexagod/repositories/work/prometheus-operator/tmp/bin/controller-gen] Error 1
```

**Review by simonpasquier — APPROVED** [review]

**Pull request merged**